XGenerate/src/main/java/com/xbreeze/xgenerate/model/ModelPreprocessor.java

```diff
@@ -24,242 +24,267 @@
 
 	/**
 	 * Pre-process the model.
+	 * 
 	 * @param model
 	 * @param modelConfig
-	 * @throws ModelPreprocessorException 
+	 * @throws ModelPreprocessorException
 	 */
 	public static void preprocessModel(Model model, ModelConfig modelConfig) throws ModelPreprocessorException {
 		logger.info("Starting model preprocessing");
 
-		/// Store the pre-processed model.
-		String preprocessedModel = model.getModelFileContent();
+		// Load the preprocessed model into memory
+		DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+		DocumentBuilder builder;
+		try {
+			builder = factory.newDocumentBuilder();
+		} catch (ParserConfigurationException exc) {			
+			throw new ModelPreprocessorException(
+					String.format("Error while reading model XML file: %s", exc.getMessage()));
+		}
+		Document doc;
+		try {
+			doc = builder.parse(new ByteArrayInputStream(model.getModelFileContent().getBytes()));
+		} catch(SAXException | IOException exc) {
+			throw new ModelPreprocessorException(
+				String.format("Error while reading model XML file: %s", exc.getMessage()));
+		}
 		
 		
 		// ModelAttributeInjections
-		// First do attribute injection, in case attributes are used as source that are removed in the next step
-		if (modelConfig != null && modelConfig.getModelAttributeInjections() != null && modelConfig.getModelAttributeInjections().size() > 0) {
-			preprocessedModel = performModelAttributeInjections(preprocessedModel, modelConfig.getModelAttributeInjections(), modelConfig.getNamespaces());
+		// First do attribute injection, in case attributes are used as source that are
+		// removed in the next step
+		if (modelConfig != null && modelConfig.getModelAttributeInjections() != null
+				&& modelConfig.getModelAttributeInjections().size() > 0) {
+			doc = performModelAttributeInjections(doc,
+					modelConfig.getModelAttributeInjections(), modelConfig.getNamespaces());
 		}
 
+		
 		// ModelNodeRemovals
-		if (modelConfig != null && modelConfig.getModelNodeRemovals() != null && modelConfig.getModelNodeRemovals().size() > 0) {
-			preprocessedModel = performModelNodeRemovals(preprocessedModel, modelConfig.getModelNodeRemovals(), modelConfig.getNamespaces());
+		if (modelConfig != null && modelConfig.getModelNodeRemovals() != null
+				&& modelConfig.getModelNodeRemovals().size() > 0) {
+			doc = performModelNodeRemovals(doc, modelConfig.getModelNodeRemovals(),
+					modelConfig.getNamespaces());
 		}
-
-		// Store the pre-processed model in the Model object.
-		model.setPreprocessedModel(preprocessedModel);
 		
+		//Transform the preprocessed document to string and store it in the model object.
+		String preprocessedModel = null;
+		DOMSource domSource = new DOMSource(doc);
+		StringWriter writer = new StringWriter();
+		StreamResult result = new StreamResult(writer);
+		TransformerFactory tf = TransformerFactory.newInstance();
+		Transformer transformer;
```

Comment:
I think this logic to transform a Document into a String should be refactored into the XmlUtils class, cause I think this will happen more often.

Reply:
Added a SaxonXMLUtils class and refactored